core/vm/state_manager.go

```diff
@@ -0,0 +1,113 @@
+package vm
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type stateManagerFunction func(*EVM, *Contract, []byte) ([]byte, error)
+type methodId [4]byte
+
+var funcs = map[string]stateManagerFunction{
+	"getStorage(address,bytes32)":            getStorage,
+	"setStorage(address,bytes32,bytes32)":    setStorage,
+	"getOvmContractNonce(address)":           getOvmContractNonce,
+	"getCodeContractBytecode(address)":       getCodeContractBytecode,
+	"getCodeContractHash(address)":           getCodeContractHash,
+	"getCodeContractAddress(address)":        getCodeContractAddress,
+	"associateCodeContract(address,address)": associateCodeContract,
+	"incrementOvmContractNonce(address)":     incrementOvmContractNonce,
+}
+var methodIds map[[4]byte]stateManagerFunction
+var executionMangerBytecode []byte
+
+func init() {
+	methodIds = make(map[[4]byte]stateManagerFunction, len(funcs))
+	for methodSignature, f := range funcs {
+		methodIds[MethodSignatureToMethodId(methodSignature)] = f
+	}
+}
+
+func MethodSignatureToMethodId(methodSignature string) [4]byte {
+	var methodId [4]byte
+	copy(methodId[:], crypto.Keccak256([]byte(methodSignature)))
+	return methodId
+}
+
+func callStateManager(input []byte, evm *EVM, contract *Contract) (ret []byte, err error) {
+	var methodId [4]byte
+	if len(input) == 0 {
+		return nil, nil
+	}
+	copy(methodId[:], input[:4])
+	ret, err = methodIds[methodId](evm, contract, input)
```

Comment:
I think this will do something weird if `methodId` is not in `methodIds`. Map lookup returns `value, exists`, where `exists` is a bool as to whether or not the value exists. Not sure what it'll do in this case? Might have to separate access and invocation? I could be wrong though.

Reply:
Hmm so I've had this error throw before and it _seemed_ to work? I should add a test for it though!